[RFR] Extract values from v3 token

kwapik opened the pull request:
For #360

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

- Roles - https://github.com/openstack/keystone/blob/32577e08b61b31cda045513f06535b474a54d2a2/keystone/models/token_model.py#L242
- Project - https://github.com/openstack/keystone/blob/32577e08b61b31cda045513f06535b474a54d2a2/keystone/models/token_model.py#L140
- User - https://github.com/openstack/keystone/blob/32577e08b61b31cda045513f06535b474a54d2a2/keystone/models/token_model.py#L89

kwapik commented:
Hi @jrperritt ,

could you take a look?
Do I need to implement some acceptance tests as well?

jtopjian commented:
@kwapik Acceptance tests aren't required but I usually end up making them at some point just to verify usability / compatibility. If you're able to write a few quick tests, go for it :)

jtopjian commented:
@kwapik In general this looks good. I have a few comments:

The code you linked to looks like the equivalent of what you're implementing here, just in Python. That's OK. I think what's also important is how the token response is being built. It looks like that is done throughout here:

https://github.com/openstack/keystone/blob/49b74dd90a5bccd03b8432978ab92a7ef4ef727c/keystone/token/providers/common.py

For example:

https://github.com/openstack/keystone/blob/49b74dd90a5bccd03b8432978ab92a7ef4ef727c/keystone/token/providers/common.py#L268

https://github.com/openstack/keystone/blob/49b74dd90a5bccd03b8432978ab92a7ef4ef727c/keystone/token/providers/common.py#L241

https://github.com/openstack/keystone/blob/49b74dd90a5bccd03b8432978ab92a7ef4ef727c/keystone/token/providers/common.py#L110

Secondly, I think the unit tests might be better if you used an entire token response as a fixture, rather than just testing small pieces of the response. Here's a fixture you can use:

https://gist.github.com/jtopjian/a7168e0418140d618e7e10ee1c80197b

Third, the example response above shows that `project` also contains `domain`. Should this be accounted for, too?

kwapik commented:
Hi @jtopjian,

Thanks for your quick review.
I wasn't sure to which part of keystone code urls should point. I'm not really familiar with keystone code itself.

> Secondly, I think the unit tests might be better if you used an entire token response as a fixture, rather than just testing small pieces of the response. Here's a fixture you can use:
>
> https://gist.github.com/jtopjian/a7168e0418140d618e7e10ee1c80197b

Updated test to use this fixture. I deleted some catalog entries, IMO there is no point to check so many of them.

> Third, the example response above shows that project also contains domain. Should this be accounted for, too?

Included this also.

> Acceptance tests aren't required but I usually end up making them at some point just to verify usability / compatibility. If you're able to write a few quick tests, go for it :)

Created simple one. I'm not sure if they are run by CI tool though.

jtopjian commented:
@kwapik Thanks!

Just one small change to the acceptance tests. Instead of:

```go
client, err := clients.NewIdentityV3UnauthenticatedClient()
```

do

```go
client, err := clients.NewIdentityV3Client()
```

I think I was originally doing the former to test authentication via a token. But that's generating an unscoped token and causing the tests to fail. The latter version makes all tests pass.

Other than that, this looks good to me and can be merged once the above is done. :)

jrperritt commented on openstack/identity/v3/tokens/results.go:
```diff
@@ -67,12 +97,37 @@ func (r commonResult) ExtractToken() (*Token, error) {
+// ExtractProject returns Project to which User is authorized.
+func (r commonResult) ExtractProject() (*Project, error) {
+	var s struct {
+		Project `json:"project"`
```
`Project *Project`

kwapik replied:
Done.

jrperritt commented on openstack/identity/v3/tokens/results.go:
```diff
@@ -67,12 +97,37 @@ func (r commonResult) ExtractToken() (*Token, error) {
+// ExtractUser returns the User that is the owner of the Token.
+func (r commonResult) ExtractUser() (*User, error) {
+	var s struct {
+		User `json:"user"`
```
`User *User`

kwapik replied:
Done.

jrperritt replied:
Not quite: this is what it should look like:
```go
User *User  `json:"user"`
```

Same with `Project` below.

kwapik replied:
Sorry, missed the second comment. Fixed it properly now. Project related change also.

jrperritt commented on openstack/identity/v3/tokens/results.go:
```diff
@@ -41,6 +41,36 @@ type ServiceCatalog struct {
+}
+
+type Roles struct {
+	Roles []*Role `json:"roles"`
```
`Roles []Role`

jtopjian replied:
Why `[]Role` here?

jrperritt replied:
1. Consistency
2. Slices are already pointers

jtopjian replied:
> Slices are already pointers

👍 

Thanks for catching the other ones, too. I'm going to see if there's a `vet` or lint gate test that can catch those.

kwapik replied:
Done.

kwapik commented:
@jrperritt @jtopjian I fixed the code according to your comments.

Side note: It is not stated in Contribution Guide how changes after review should be commited. I created `Fixup` commits and will rebase them after approval, but I could also amend previous ones or create new like `Review changes`. Personally I prefer the first approach, because it's outcome is the cleaner during review and after merge. Which one do you prefer and recommend in this repo?

jtopjian commented:
> Personally I prefer the first approach, because it's outcome is the cleaner during review and after merge. Which one do you prefer and recommend in this repo?

That's the correct approach. You also don't need to squash or rebase prior to a merge. Github can handle that now. 

The STYLEGUIDE has the following note:

> Unless explicitly asked, do not squash commits in the middle of a review; only append. It makes it difficult for the reviewer to see what's changed from one review to the next.

Although maybe it should be moved / copied to the CONTRIBUTING guide.

kwapik commented:
@jrperritt 
code should be ok now.

jrperritt commented on openstack/identity/v3/tokens/results.go:
```diff
@@ -67,12 +97,37 @@ func (r commonResult) ExtractToken() (*Token, error) {
+}
+
+// ExtractRoles returns Roles to which User is authorized.
+func (r commonResult) ExtractRoles() (*Roles, error) {
```
Instead of `*Roles`, let's just return `[]Role` and remove the `Roles` type.

kwapik replied:
@jrperritt Done.

BTW I probably messed something up while creating fixups and they won't squash automatically. This would need manual work.

kwapik commented:
@jrperritt could you take a look one more time?

jrperritt commented:
LGTM
